lib/minitest/assertions.rb

```diff
@@ -464,25 +464,31 @@ def capture_io
     def capture_subprocess_io
       _synchronize do
         begin
-          require "tempfile"
-
-          captured_stdout, captured_stderr = Tempfile.new("out"), Tempfile.new("err")
+          stdout_reader, stdout_writer = IO.pipe
+          stderr_reader, stderr_writer = IO.pipe
 
           orig_stdout, orig_stderr = $stdout.dup, $stderr.dup
-          $stdout.reopen captured_stdout
-          $stderr.reopen captured_stderr
 
-          yield
+          begin
+            $stdout.reopen stdout_writer
+            $stderr.reopen stderr_writer
```

Comment:
This is more of a theoretical issue, but `$stderr` and `$stdout` are "normal" Ruby variables that the user could have overwritten to something else. It might be better to use the constants instead, `STDERR` and `STDOUT`.

Reply:
Exactly the opposite. `STDERR` and `STDOUT` are constants are really only there so you can restore `$stderr` and `$stdout`. `puts` and `warn` and friends use the globals to output and so should we. If the user overwrote them to something else, it was probably for a reason and we still want to be able to test that.

Reply:
> If the user overwrote them to something else, it was probably for a reason and we still want to be able to test that

But a subprocess will not pick up on that. So if the user assigned something else to these variables in the parent process, this will have no effect on the subprocess.

Reply:
But... You have that problem anyways. Using the constants won't help if they're not being used.

Reply:
I wrote this test script
``` ruby
require 'minitest'

class Foo
  include Minitest::Assertions

  def call
    stdout, _ = capture_subprocess_io do
      system("echo From subprocess")
      $stdout.puts "From parent using $stdout"
      STDOUT.puts "From parent using STDOUT"
    end
  end
end

output, _ = Foo.new.call

STDERR.puts "> Captured string before redirecting $stdout:"
STDERR.print output

$stdout = Tempfile.new
begin
  output, _ = Foo.new.call

  STDERR.puts "> Captured string after redirect:"
  STDERR.print output
ensure
  $stdout.unlink
  $stdout.close
end
```

When I run this, using the current implementation (which uses $stdout/$stderr):

```
> Captured string before redirecting $stdout:
From subprocess
From parent using $stdout
From parent using STDOUT
From subprocess
From parent using STDOUT
> Captured string after redirect:
From parent using $stdout
```

When I switch the implementation to use STDOUT and STDERR instead:
```
> Captured string before redirecting $stdout:
From subprocess
From parent using $stdout
From parent using STDOUT
> Captured string after redirect:
From subprocess
From parent using STDOUT
```

Reply:
There is no perfect solution to this problem. I prefer the second option, because it will at least capture the IO from the subprocess, which is what the name of the method implies.

Note that this is also an issue with the current implementation using tempfiles.